snowflake/connect.go

```diff
@@ -0,0 +1,277 @@
+		config.Token = oauthAccessToken
+	} else if password != "" {
+		config.Password = password
+	} else {
+		return "", errors.New("no authentication method provided")
+	}
+
+	return gosnowflake.DSN(&config)
+}
+
+func ReadPrivateKeyFile(privateKeyPath string) ([]byte, error) {
+	expandedPrivateKeyPath, err := homedir.Expand(privateKeyPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "Invalid Path to private key")
+	}
+
+	privateKeyBytes, err := ioutil.ReadFile(expandedPrivateKeyPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "Could not read private key")
+	}
+
+	if len(privateKeyBytes) == 0 {
+		return nil, errors.New("Private key is empty")
+	}
+
+	return privateKeyBytes, nil
+}
+
+func ParsePrivateKey(privateKeyBytes []byte, passhrase []byte) (*rsa.PrivateKey, error) {
+	privateKeyBlock, _ := pem.Decode(privateKeyBytes)
+	if privateKeyBlock == nil {
+		return nil, fmt.Errorf("Could not parse private key, key is not in PEM format")
+	}
+
+	if privateKeyBlock.Type == "ENCRYPTED PRIVATE KEY" {
+		if len(passhrase) == 0 {
+			return nil, fmt.Errorf("Private key requires a passphrase, but private_key_passphrase was not supplied")
+		}
+		privateKey, err := pkcs8.ParsePKCS8PrivateKeyRSA(privateKeyBlock.Bytes, passhrase)
+		if err != nil {
+			return nil, errors.Wrap(
+				err,
+				"Could not parse encrypted private key with passphrase, only ciphers aes-128-cbc, aes-128-gcm, aes-192-cbc, aes-192-gcm, aes-256-cbc, aes-256-gcm, and des-ede3-cbc are supported",
+			)
+		}
+		return privateKey, nil
+	}
+
+	privateKey, err := ssh.ParseRawPrivateKey(privateKeyBytes)
+	if err != nil {
+		return nil, errors.Wrap(err, "Could not parse private key")
+	}
+
+	rsaPrivateKey, ok := privateKey.(*rsa.PrivateKey)
+	if !ok {
+		return nil, errors.New("privateKey not of type RSA")
+	}
+	return rsaPrivateKey, nil
+}
+
+type Result struct {
+	AccessToken string `json:"access_token"`
+	TokenType   string `json:"token_type"`
+	ExpiresIn   int    `json:"expires_in"`
+}
+
+func GetOauthData(refreshToken, redirectUrl string) url.Values {
+	data := url.Values{}
+	data.Set("grant_type", "refresh_token")
+	data.Set("refresh_token", refreshToken)
+	data.Set("redirect_uri", redirectUrl)
+	return data
+}
+
+func GetOauthRequest(dataContent io.Reader, endPoint, clientId, clientSecret string) (*http.Request, error) {
+	request, err := http.NewRequest("POST", endPoint, dataContent)
+	if err != nil {
+		return nil, errors.Wrap(err, "Request to the endpoint could not be completed")
+	}
+	request.SetBasicAuth(clientId, clientSecret)
+	request.Header.Set("Content-Type", "application/x-www-form-urlencoded;charset=UTF-8")
+	return request, nil
+
+}
+
+func GetOauthAccessToken(
+	endPoint,
+	client_id,
+	client_secret string,
+	data url.Values) (string, error) {
+
+	client := &http.Client{}
+	request, err := GetOauthRequest(strings.NewReader(data.Encode()), endPoint, client_id, client_secret)
+	if err != nil {
+		return "", errors.Wrap(err, "Oauth request returned an error:")
```

Comment:
What's the difference between using `errors.Wrap`, `fmt.Errorf`, and `errors.New`? How is the output different? 

From what I could find online, the `errors` package has been archived/deprecated (https://github.com/pkg/errors), so it seems like using `fmt.Errorf` with `%w` is better. We may lose the stack trace at the time, but I'm not sure if we need the full stack trace anwyay?

Reply:
After a quick discussion, we decided to use `fmt.Errorf` for error handling in this plugin (and others)